_inc/lib/class.core-rest-api-endpoints.php

```diff
@@ -2956,6 +2985,111 @@ private static function core_get_plugins() {
 		return array();
 	}
 
+
+	/**
+	 * Get third party plugin API keys.
+	 *
+	 * @param WP_REST_Request $request {
+	 *     Array of parameters received by request.
+	 *
+	 *     @type string $slug Plugin slug with the syntax 'plugin-directory/plugin-main-file.php'.
+	 * }
+	 */
+	public static function get_api_key( $request ) {
+		$service = $request['service'];
+		$option  = self::key_for_api_service( $service );
+		return array(
+			'api_key' => Jetpack_Options::get_option( $option, '' ),
+		);
+	}
+
+	/**
+	 * Update third party plugin API keys.
+	 *
+	 * @param WP_REST_Request $request {
+	 *     Array of parameters received by request.
+	 *
+	 *     @type string $slug Plugin slug with the syntax 'plugin-directory/plugin-main-file.php'.
+	 * }
+	 */
+	public static function update_api_key( $request ) {
+		$params     = $request->get_json_params();
+		$service    = $request['service'];
+		$api_key    = trim( $params['api_key'] );
+		$option     = self::key_for_api_service( $service );
+		$validation = self::validate_api_key( $api_key, $service );
+
+		if ( ! $validation['status'] ) {
+			return new WP_Error( 'invalid_key', esc_html__( 'Invalid API Key', 'jetpack' ), array( 'status' => 404 ) );
+		}
+
+		Jetpack_Options::update_option( $option, $api_key );
+		return array(
+			'api_key' => Jetpack_Options::get_option( $option, '' ),
+			'message' => esc_html__( 'API key set successfully!', 'jetpack' ),
+		);
+	}
+
+	/**
+	 * Delete a third party plugin API key.
+	 *
+	 * @param WP_REST_Request $request {
+	 *     Array of parameters received by request.
+	 *
+	 *     @type string $slug Plugin slug with the syntax 'plugin-directory/plugin-main-file.php'.
+	 * }
+	 */
+	public static function delete_api_key( $request ) {
+		$service = $request['service'];
+		$option  = self::key_for_api_service( $service );
+		Jetpack_Options::delete_option( $option );
+		return array(
+			'api_key' => Jetpack_Options::get_option( $option, '' ),
+			'message' => esc_html__( 'API key deleted successfully!', 'jetpack' ),
+		);
+	}
+
+	/**
+	 * Validate API Key
+	 *
+	 * @param string $key The API key to be validated.
+	 * @param string $service The service the API key is for.
+	 */
+	public static function validate_api_key( $key = null, $service = null ) {
+		$validation = false;
+		switch ( $service ) {
+			case 'googlemaps':
+				$validation = self::validate_api_key_googlemaps( $key );
+				break;
+		}
+		return $validation;
+	}
+
+	/**
+	 * Validate Google Maps API key
+	 *
+	 * @param string $key The API key to be validated.
+	 */
+	public static function validate_api_key_googlemaps( $key = null ) {
+		$address = '1+broadway+new+york+ny+usa';
+		$path    = 'https://maps.googleapis.com/maps/api/place/findplacefromtext/json?inputtype=textquery&input=' . $address;
+		$path    = add_query_arg( 'key', $key, $path );
+		$json    = json_decode( wp_remote_retrieve_body( wp_remote_get( esc_url( $path, null, null ) ) ), true );
```

Comment:
`wp_remote_get( esc_url( $path, null, null ) )`  can sometimes return a wp_error object. 
So it is a good practice to check for it  before passing that the result to `wp_remote_retrieve_body( )`

```
$result = wp_remote_get( esc_url( $path, null, null ) );
if ( is_wp_error( $result ) ) {
   return array( 'status' => false, 'error_message' => $result->error_message() ) //
}
$json = json_decode( wp_remote_retrieve_body( $result ), true );
```

Reply:
This should probably be `esc_url_raw()`. `esc_url()` is essentially: `HTML_encode( esc_url_raw() )`.

Reply:
We should also be using `wp_safe_remote_get()`.

Reply:
`wp_remote_get()` error catch, `esc_url_raw()` and `wp_safe_remote_get()` all resolved in https://github.com/Automattic/jetpack/pull/10393/commits/8274b5e211418496252ea6a58c6882d06c9c5ba9